### PR TITLE
add SubscribePaymentAccepted

### PR DESCRIPTION
**Pull request by joostjager**
If invoices are truly stateless not just for lnmux, but also for services that use lnmux, it isn't possible to use `SubscribeSingleInvoice` with a hash.

This PR adds a new rpc `SubscribePaymentAccepted` that sends out an event for every payment that reaches the accepted state.

### CODE REVIEW

File: cmd/lnmuxd/lnmux_proto/lnmux.proto [hunk 1]
```diff
@@ -9,6 +9,8 @@ service Service {
 
     rpc SubscribeSingleInvoice (SubscribeSingleInvoiceRequest) returns (stream SubscribeSingleInvoiceResponse);
 
+    rpc SubscribePaymentAccepted (SubscribePaymentAcceptedRequest) returns (stream SubscribePaymentAcceptedResponse);
```
Comment (aakselrod-nydig) [inline cmd/lnmuxd/lnmux_proto/lnmux.proto:12]: ⚡

File: cmd/lnmuxd/lnmux_proto/lnmux.proto [hunk 2]
```diff
@@ -41,6 +43,12 @@ message SubscribeSingleInvoiceResponse {
     InvoiceState state = 1;
 }
 
+message SubscribePaymentAcceptedRequest {}
+
+message SubscribePaymentAcceptedResponse {
+    bytes hash = 1;
+}
+
```
Comment (aakselrod-nydig) [inline cmd/lnmuxd/lnmux_proto/lnmux.proto:51]: ⚡ 🔥

File: cmd/lnmuxd/server.go [hunk 3]
```diff
@@ -61,34 +61,50 @@ func (s *server) SubscribeSingleInvoice(req *lnmux_proto.SubscribeSingleInvoiceR
...
+		},
+		func(update lnmux.InvoiceUpdate) error {
+			return subscription.Send(&lnmux_proto.SubscribeSingleInvoiceResponse{
+				State: marshallInvoiceState(update.State),
+			})
+		},
+	)
+}
+
+func streamBuffered[T any](ctx context.Context,
```
Comment (aakselrod-nydig) [inline cmd/lnmuxd/server.go:77]: ⚡ 🔥 ⚡ 
great place to use generics! will save lots of work for adding to the API in the future.

File: invoiceregistry.go [hunk 4]
```diff
@@ -148,17 +152,19 @@ func NewRegistry(cdb *persistence.PostgresPersister,
...
-		quit:                      make(chan struct{}),
+		cdb:                             cdb,
+		hodlSubscriptions:               make(map[types.CircuitKey][]func(HtlcResolution)),
+		cfg:                             cfg,
+		invoices:                        make(map[lntypes.Hash]*invoiceState),
+		htlcChan:                        make(chan *registryHtlc),
+		newInvoiceSubscription:          make(chan invoiceSubscription),
+		cancelInvoiceSubscription:       make(chan int),
+		newInvoiceAcceptSubscription:    make(chan invoiceAcceptSubscription),
+		cancelInvoiceAcceptSubscription: make(chan int),
```
Comment (aakselrod-nydig) [inline invoiceregistry.go:163]: ⚡

File: invoiceregistry.go [hunk 5]
```diff
@@ -205,10 +211,35 @@ func (i *InvoiceRegistry) Subscribe(hash lntypes.Hash,
...
-			id:   subscriberId,
-			hash: hash,
-		}:
+		case i.cancelInvoiceSubscription <- subscriberId:
+		case <-i.quit:
+		}
+	}, nil
+}
+
+func (i *InvoiceRegistry) SubscribeAccept(callback AcceptCallback) (
```
Comment (aakselrod-nydig) [inline invoiceregistry.go:220]: ⚡

File: invoiceregistry.go [hunk 6]
```diff
@@ -471,6 +511,27 @@ func (i *InvoiceRegistry) addSubscriber(ctx context.Context,
 	return nil
 }
 
+func (i *InvoiceRegistry) addAcceptSubscriber(ctx context.Context,
```
Comment (aakselrod-nydig) [inline invoiceregistry.go:514]: ⚡

File: invoiceregistry_test.go [hunk 7]
```diff
@@ -120,6 +120,16 @@ func (r *registryTestContext) subscribe(hash lntypes.Hash) (chan InvoiceUpdate,
 	return updateChan, cancel
 }
 
+func (r *registryTestContext) subscribeAccept() (chan lntypes.Hash, func()) {
```
Comment (aakselrod-nydig) [inline invoiceregistry_test.go:123]: ⚡

File: invoiceregistry_test.go [hunk 8]
```diff
@@ -162,6 +172,7 @@ func TestAutoSettle(t *testing.T) {
 
 	// Subscribe to updates for invoice.
 	updateChan, cancelUpdates := c.subscribe(preimage.Hash())
+	acceptChan, cancelAcceptUpdates := c.subscribeAccept()
```
Comment (aakselrod-nydig) [inline invoiceregistry_test.go:175]: ⚡

File: mux_test.go [hunk 9]
```diff
@@ -147,6 +148,21 @@ func TestMux(t *testing.T) {
...
+	})
+	require.NoError(t, err)
+
+	expectAccept := func(expectedHash lntypes.Hash) {
+		t.Helper()
+
+		hash := <-acceptChan
+		require.Equal(t, expectedHash, hash)
+	}
+
```
Comment (aakselrod-nydig) [inline mux_test.go:165]: ⚡ this is a really good example of how to use the new subscription type

File: invoiceregistry.go [hunk 10]
```diff
@@ -471,6 +511,27 @@ func (i *InvoiceRegistry) addSubscriber(ctx context.Context,
...
+
+	// Store subscriber for future updates.
+	i.subscriptionManager.addAcceptSubscription(
+		newSubscription.id, newSubscription.callback,
+	)
+
+	// Send accepted invoices from memory.
+	for hash, invoiceState := range i.invoices {
+		// No event for partially accepted invoices.
+		if !invoiceState.isSetComplete() {
```
Comment (aakselrod-nydig) [inline invoiceregistry.go:525]: ⚡ 🔥 ⚡

File: subscription_manager.go [hunk 11]
```diff
@@ -3,41 +3,53 @@ package lnmux
...
 	"github.com/lightningnetwork/lnd/lntypes"
 	"go.uber.org/zap"
 )
 
 type subscriptionManager struct {
-	subscriptions map[lntypes.Hash]map[int]InvoiceCallback
-	logger        *zap.SugaredLogger
+	subscriptions       map[lntypes.Hash]map[int]InvoiceCallback
+	subscriptionHash    map[int]lntypes.Hash
+	acceptSubscriptions map[int]AcceptCallback
```
Comment (aakselrod-nydig) [inline subscription_manager.go:14]: ⚡

File: invoiceregistry.go [hunk 12]
```diff
@@ -325,8 +356,17 @@ func (i *InvoiceRegistry) invoiceEventLoop(ctx context.Context) error {
...
-		case request := <-i.cancelInvoiceSubscription:
-			i.subscriptionManager.deleteSubscription(request.hash, request.id)
+		case newAcceptSubscription := <-i.newInvoiceAcceptSubscription:
+			err := i.addAcceptSubscriber(ctx, newAcceptSubscription)
+			if err != nil {
+				return err
+			}
+
+		case id := <-i.cancelInvoiceSubscription:
+			i.subscriptionManager.deleteSubscription(id)
```
Comment (aakselrod-nydig) [inline invoiceregistry.go:366]: 🔥

File: invoiceregistry.go [hunk 13]
```diff
@@ -325,8 +356,17 @@ func (i *InvoiceRegistry) invoiceEventLoop(ctx context.Context) error {
...
+
+		case id := <-i.cancelInvoiceAcceptSubscription:
+			i.subscriptionManager.deleteAcceptSubscription(id)
```
Comment (aakselrod-nydig) [inline invoiceregistry.go:369]: ⚡

File: subscription_manager.go [hunk 14]
```diff
@@ -3,41 +3,53 @@ package lnmux
...
+		subscriptions:       make(map[lntypes.Hash]map[int]InvoiceCallback),
+		subscriptionHash:    make(map[int]lntypes.Hash),
+		acceptSubscriptions: make(map[int]AcceptCallback),
+		logger:              logger,
 	}
 }
 
-func (s *subscriptionManager) deleteSubscription(hash lntypes.Hash, id int) {
-	subs, ok := s.subscriptions[hash]
+func (s *subscriptionManager) deleteSubscription(id int) {
```
Comment (aakselrod-nydig) [inline subscription_manager.go:29]: ⚡ 🔥 ⚡ this is a really nice change

File: subscription_manager.go [hunk 15]
```diff
@@ -3,41 +3,53 @@ package lnmux
...
-		s.logger.Debugw("subscription not found",
-			"id", id, "hash", hash)
+	subs, ok := s.subscriptions[hash]
+	if !ok {
+		panic("inconsistent subscription")
+	}
 
-		return
+	if _, ok := subs[id]; !ok {
+		panic("inconsistent subscription")
```
Comment (aakselrod-nydig) [inline subscription_manager.go:44]: 👍

File: subscription_manager.go [hunk 16]
```diff
@@ -48,6 +60,26 @@ func (s *subscriptionManager) addSubscription(hash lntypes.Hash, id int,
 		subs[hash] = make(map[int]InvoiceCallback)
 	}
 	subs[hash][id] = callback
+
+	s.subscriptionHash[id] = hash
+}
+
+func (s *subscriptionManager) addAcceptSubscription(id int,
```
Comment (aakselrod-nydig) [inline subscription_manager.go:67]: ⚡

File: subscription_manager.go [hunk 17]
```diff
@@ -48,6 +60,26 @@ func (s *subscriptionManager) addSubscription(hash lntypes.Hash, id int,
...
+	callback AcceptCallback) {
+
+	s.acceptSubscriptions[id] = callback
+}
+
+func (s *subscriptionManager) deleteAcceptSubscription(id int) {
+	_, ok := s.acceptSubscriptions[id]
+	if !ok {
+		s.logger.Debugw("no accept subnscription",
```
Comment (aakselrod-nydig) [inline subscription_manager.go]: nit: typo

File: subscription_manager.go [hunk 18]
```diff
@@ -48,6 +60,26 @@ func (s *subscriptionManager) addSubscription(hash lntypes.Hash, id int,
...
+	s.subscriptionHash[id] = hash
+}
+
+func (s *subscriptionManager) addAcceptSubscription(id int,
+	callback AcceptCallback) {
+
+	s.acceptSubscriptions[id] = callback
+}
+
+func (s *subscriptionManager) deleteAcceptSubscription(id int) {
```
Comment (aakselrod-nydig) [inline subscription_manager.go:73]: ⚡

File: subscription_manager.go [hunk 19]
```diff
@@ -57,6 +89,12 @@ func (s *subscriptionManager) generateSubscriptionId() int {
 func (s *subscriptionManager) notifySubscribers(hash lntypes.Hash,
 	update InvoiceUpdate) {
 
+	if update.State == persistence.InvoiceStateAccepted {
+		for _, subscriber := range s.acceptSubscriptions {
+			subscriber(hash)
+		}
+	}
+
```
Comment (aakselrod-nydig) [inline subscription_manager.go:96]: ⚡

**Review by aakselrod-nydig — APPROVED** [review]
This looks great!

File: subscription_manager.go (hunk 17 above)
Comment (joostjager) [reply subscription_manager.go]: fixed

**Pull request merged**